k9mail/src/main/java/com/fsck/k9/account/AndroidAccountOAuth2TokenStore.java

```diff
@@ -0,0 +1,136 @@
+package com.fsck.k9.account;
+
+import android.accounts.Account;
+import android.accounts.AccountManager;
+import android.accounts.AccountManagerCallback;
+import android.accounts.AccountManagerFuture;
+import android.accounts.AuthenticatorException;
+import android.accounts.OperationCanceledException;
+import android.app.Activity;
+import android.content.Context;
+import android.os.Bundle;
+
+import com.fsck.k9.R;
+import com.fsck.k9.mail.AuthenticationFailedException;
+import com.fsck.k9.mail.oauth.AuthorizationException;
+import com.fsck.k9.mail.oauth.OAuth2TokenProvider;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * An interface between the OAuth2 requirements used for authentication and the AccountManager.
+ */
+public class AndroidAccountOAuth2TokenStore implements OAuth2TokenProvider {
+    private static final String GMAIL_AUTH_TOKEN_TYPE = "oauth2:https://mail.google.com/";
+    private static final String GOOGLE_ACCOUNT_TYPE = "com.google";
+
+    private Map<String,String> authTokens = new HashMap<>();
+    private AccountManager accountManager;
+
+    public AndroidAccountOAuth2TokenStore(Context applicationContext) {
+        this.accountManager = AccountManager.get(applicationContext);
+    }
+
+    @Override
+    public void authorizeAPI(final String emailAddress, final Activity activity,
+                             final OAuth2TokenProviderAuthCallback callback) {
+        Account account = getAccountFromManager(emailAddress);
+        if (account == null) {
+            callback.failure(new AuthorizationException(activity.getString(R.string.xoauth2_account_doesnt_exist)));
+            return;
+        }
+        if(account.name.equals(emailAddress)) {
+            accountManager.getAuthToken(account, GMAIL_AUTH_TOKEN_TYPE, null, activity,
+                new AccountManagerCallback<Bundle>() {
+                    @Override
+                    public void run(AccountManagerFuture<Bundle> future) {
```

Comment:
this method has borderline complexity for its nesting depth.. given its if/else/ends-in-callback structure though I don't think the code would be more readable otherwise, so might as well leave it as is